proxy/src/main/java/com/velocitypowered/proxy/protocol/netty/GS4QueryHandler.java

```diff
@@ -0,0 +1,189 @@
+package com.velocitypowered.proxy.protocol.netty;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.velocitypowered.api.proxy.Player;
+import com.velocitypowered.proxy.VelocityServer;
+import com.velocitypowered.proxy.protocol.ProtocolConstants;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.channel.socket.DatagramPacket;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.net.InetAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+public class GS4QueryHandler extends SimpleChannelInboundHandler<DatagramPacket> {
+    private static final Logger logger = LogManager.getLogger(GS4QueryHandler.class);
+
+    private final static short QUERY_MAGIC_FIRST = 0xFE;
+    private final static short QUERY_MAGIC_SECOND = 0xFD;
+    private final static byte QUERY_TYPE_HANDSHAKE = 0x09;
+    private final static byte QUERY_TYPE_STAT = 0x00;
+    private final static byte[] QUERY_RESPONSE_FULL_PADDING = new byte[] { 0x73, 0x70, 0x6C, 0x69, 0x74, 0x6E, 0x75, 0x6D, 0x00, (byte) 0x80, 0x00 };
+    private final static byte[] QUERY_RESPONSE_FULL_PADDING2 = new byte[] { 0x01, 0x70, 0x6C, 0x61, 0x79, 0x65, 0x72, 0x5F, 0x00, 0x00 };
+
+    // Contents to add into basic stat response. See ResponseWriter class below
+    private final static List<String> QUERY_BASIC_RESPONSE_CONTENTS = Arrays.asList(
```

Comment:
Suggest using `ImmutableList#of()` here.

Reply:
Tbh ImmutableSet would be even better choice here.

Reply:
That would work fine, too.